src/ParametersHandler/YarpImplementation/src/YarpImplementation.cpp

```diff
@@ -114,49 +114,136 @@ YarpImplementation::YarpImplementation(const yarp::os::Searchable& searchable)
+    // Validate the Searchable before processing it
+    for (size_t i = 0; i < bottle.size(); ++i)
     {
-        yarp::os::Value& bb = bot.get(i);
-
-        yarp::os::Bottle* sub = bb.asList();
-        if ((sub) && (sub->size() > 1))
+        if (!bottle.get(i).isList())
         {
-            std::string name = sub->get(0).toString();
-            yarp::os::Bottle* subSub = sub->get(1).asList();
-            if ((subSub) && (subSub->size() > 1))
-            {
-                m_lists.emplace(name, std::make_shared<YarpImplementation>(*sub));
-            } else
-            {
-                m_container.add(bb);
-            }
-
-        } else
+            std::cerr << "Searchable not valid: " << bottle.get(i).toString() << std::endl;
+            return;
+        }
+    }
+
+    // Clear the handler
+    clear();
+
+    enum class Type
+    {
+        GROUP,
+        SCALAR,
+        VECTOR,
+    };
+
+    // Loop over pairs:
+    // 1. (name, scalar_value)
+    // 2. (name, vector_value)
+    // 3. (name, group_value1, group_value2, ...)
+    for (size_t i = 0; i < bottle.size(); ++i)
+    {
+        // Get the pair
+        const yarp::os::Value& pair = bottle.get(i);
+
+        // Parse the pair into (name, value)
+        std::string name;
+        yarp::os::Value value;
+
+        const auto type = [&pair, &name, &value]() {
+            // Unpack the pair
+            auto* list = pair.asList();
+            // const std::string name = list->get(0).asString();
+            // const yarp::os::Value& value = list->get(1);
+            name = list->get(0).asString();
+            value = list->get(1);
+
+            // If there are more than 2 elements is for sure a group
+            if (pair.asList()->size() > 2)
+                return Type::GROUP;
+
+            // If the value is not a list, it is a scalar parameter
+            if (!value.isList())
+                return Type::SCALAR;
+
+            // If the value is a list, is a vector parameter.
+            // Let's leave out the case of two elements.
+            if (value.asList()->size() != 2)
+                return Type::VECTOR;
+
+            // Now it could be either a group with one pair of parameters or
+            // a vector parameter with 2 elements:
+            //      GROUP_NAME (param1 value1)
+            //      param_name (elem1 elem2)
+
+            // If the first element of the value is not a string, it cannot be a group
+            if (!value.asList()->get(0).isString())
+                return Type::VECTOR;
+
+            // This last case is the most problematic because we cannot be sure.
+            // Let's decide that it is a group if the name is uppercase.
+            if (std::all_of(name.begin(), name.end(), [](unsigned char c) {
+                    return std::isupper(c) || c == '_' || c == '-';
+                }))
+                return Type::GROUP;
```

Comment:
This is the problematic part, and the reason why I used ``lists`` and not ``groups`` in the name of the class member in the first place. 
To be honest, I don't feel very comfortable in adding this kind of constraint since it would be implementation-dependent. The rule would be, 
"If the group contains only two elements, its name has to be all uppercase, while a vector of two elements cannot have an uppercase name, but only when using the YarpImplementation". 

In addition, what happens if the name of the group is or contains a number?

Another possible case, that right now is not considered in the tests, is the presence of flags (i.e. parameters with no value, like the deprecated ``verbose``). Before, they would have been simply included, but not lost, while they are right now.

Reply:
I expected that to be a problem, and unfortunately I didn't find any better solution. The limitation is that YARP does not offer any way I know to check the problematic case you quoted.

> To be honest, I don't feel very comfortable in adding this kind of constraint since it would be implementation-dependent. The rule would be, "If the group contains only two elements, its name has to be all uppercase, while a vector of two elements cannot have an uppercase name, but only when using the YarpImplementation".

I understand your point and your concerns. I hope that in the future, when the new toml-based configuration will be finalized, those classes will be polished and their behaviour improved (maybe @traversaro knows more). What I want to remark is that this piece of code is specific to the `YarpImplementation`, and this exception could be properly documented. Numbers can of course be added to the allowed characters for groups.

This being said, I am aware that this feature became big for this PR and I really hoped to find a better workaround with less LOC. Unfortunately many limitations prevented that.

 > Another possible case, that right now is not considered in the tests, is the presence of flags (i.e. parameters with no value, like the deprecated verbose). Before, they would have been simply included, but not lost, while they are right now.

Is this documented anywhere? I couldn't find any reference of flags. Note that flags implemented like you said can be mapped to boolean scalar parameters, and their semantics would not change (and probably the implementations would benefit from this unified approach).

I close this comment saying that as new user, I find the possibility to list the names of the groups and the parameters very important. Maybe this time it will not go through, but I suggest that is better to handle it now that there aren't many blf users and the APIs are not yet frozen. To me, if properly documented, the problematic case is acceptable, not nice, not optimal, but acceptable. And it would be a pity preventing the integration of simple features like this due to one YARP limitation that can be lift in the future.

Reply:
I understand, [I have been trough this already](https://github.com/dic-iit/bipedal-locomotion-framework/pull/18#issue-378217180). I just don't feel comfortable in adding a constraint (that is basically a regression) for a feature that is somehow comparable to [``toString()``](https://github.com/dic-iit/bipedal-locomotion-framework/blob/master/src/ParametersHandler/include/BipedalLocomotion/ParametersHandler/IParametersHandler.h#L193).
In the end, I believe we can obtain a good trade-off by printing parameters and group names together. This would also avoid catching exceptions in the ``std`` implementation. We just need to find a decent name for the method (like ``getTags``). 
Would this (printing all together) be a reasonable way out?

Another possibility would be to have the output divided in (estimated) categories. This would be reasonable since the internals don't change, and we can clearly state that the output categorization may not be accurate for some corner cases. 

Yet another possibility may be a sum of the two previous. We keep both methods, the categorization is only done in the method to get the names (in the end you just need to deal with the lists, because all other parameters have already been categorized), with the statement of possible inaccuracies. And if we want, we add also a method to get them all, free of any possible misinterpretations. 

What do you think?

Reply:
> I just don't feel comfortable in adding a constraint (that is basically a regression) for a feature that is somehow comparable to toString().

If I don't mistake, from this method is not possible to understand if an option is a group or list, right? In the long run it would be nice having a uniform and more structured output for all the implementations, but I fear this would require the changes we're discussing. A uniform serialization allows to build new objects from string, and is the base of an IDL system. It can be useful for a parameter server or similar solutions.

> I understand, I have been through this already.

I'm sure you know the frustration when you have to deal with Bottle and Values :)

> Another possibility would be to have the output divided in (estimated) categories. This would be reasonable since the internals don't change, and we can clearly state that the output categorization may not be accurate for some corner cases.

Mmh between this solution and the problematic case under discussion, I would rather prefer keeping the problematic case. One thing is being warned that a rather uncommon and well defined corner case can happen , another is reading some output with the impossibility to know if it is accurate or not. I say uncommon because the group names of YARP files are typically upper case, and this exception happens only when such group has only one parameter that is a list or two strings. 

From the software design point of view, I am also a bit concerned by the fact that the introspection of this implementation is just impossible, and is left entirely to the user.

> Yet another possibility may be a sum of the two previous. We keep both methods, the categorization is only done in the method to get the names (in the end you just need to deal with the lists, because all other parameters have already been categorized), with the statement of possible inaccuracies. And if we want, we add also a method to get them all, free of any possible misinterpretations.
>
> What do you think?

I'm not sure I got it entirely :) You suggest to keep the two methods to get group names and parameter names, that provide a wrong output (because inaccurate=wrong) in few cases, and also add the combination of both that is always correct. 

---

I have another idea.

Keep the new implementation I proposed and store the problematic parameter both as a group and as a list. Then, it's the user that decides how to gather it, and in both cases they would not be surprised. This solution would provide a less "inaccurate" output for the methods that return names because it would show a group and a list with the same name and the same content. That is also the current situation of the underlying storage. This way, we solve gracefully the corner case without compromising any functionality.

Reply:
>Keep the new implementation I proposed and store the problematic parameter both as a group and as a list. Then, it's the user that decides how to gather it, and in both cases they would not be surprised. This solution would provide a less "inaccurate" output for the methods that return names because it would show a group and a list with the same name and the same content. That is also the current situation of the underlying storage. This way, we solve gracefully the corner case without compromising any functionality.

This would create duplication, which is bad since at a certain point they may diverge because of the ``set*`` method.

> You suggest to keep the two methods to get group names and parameter names, that provide a wrong output (because inaccurate=wrong) in few cases, and also add the combination of both that is always correct.

Yes, exactly. Notice that the cases that are inaccurate (see "wrong") are the same cases of your implementation. You can still perform the same assumptions on the naming, but at least, if the user doesn't care about getting the names in the correct category (i.e. group or parameter), he can still use the handler as it had in mind (with all the group names lowercase, for example), without having to remember a corner case rule. In other words, what I am proposing, is to move all the logic you put in the parsing, in the method to get the names. I am trading a constraint for the user with a line of documentation.

Reply:
> This would create duplication, which is bad since at a certain point they may diverge because of the set* method.

Yes, but the user that expects to have a group has a group, and the user that expects to have a list of strings has a list of strings.

> Yes, exactly. Notice that the cases that are inaccurate (see "wrong") are the same cases of your implementation. You can still perform the same assumptions on the naming, but at least, if the user doesn't care about getting the names in the correct category (i.e. group or parameter), he can still use the handler as it had in mind (with all the group names lowercase, for example), without having to remember a corner case rule. In other words, what I am proposing, is to move all the logic you put in the parsing, in the method to get the names. I am trading a constraint for the user with a line of documentation.

I see your point, I still believe that the duplication is the best way to handle this case, but I will follow you comment and change the implementation as you described (in a new PR).

Reply:
Great! So, what would be your plan for this PR? :grin:

Reply:
I did not follow this discussion to get the problem, but if there is some limitation in YARP that you think could be addressed, I think we can open an issue upstream.

Reply:
>I did not follow this discussion to get the problem, but if there is some limitation in YARP that you think could be addressed, I think we can open an issue upstream.

To summarize, from a ``yarp`` ``bottle`` it is impossible to distinguish a vector from a group in some corner cases (see [here](https://github.com/dic-iit/bipedal-locomotion-framework/pull/18#issue-378217180)). This is problematic with the methods ``getParameterNames`` and  ``getGroupNames``, introduced with this PR. In particular, for some entries stored in the underlying ``Bottle``, it is not possible to understand if it is a parameter or a group beforehand. 

What @diegoferigo suggested (sorry if I got it wrong) was to refactor the parsing logic of an input ``Searchable`` in order to divide groups from parameters at the source, and, for the corner cases, require the group names to be all uppercase. 

From my side, I preferred not to change the parsing logic, preserving the input structure as much as possible. In this way the user does not have to know about the uppercase rule that is valid only for the ``yarp`` implementation. The same result of what @diegoferigo suggested, nonetheless, can be achieved by replaying the above parsing logic directly inside the ``getParameterNames`` and  ``getGroupNames`` methods (with the drawback of adding logic in some getters).

I don't know if this can be addressed in anyway upstream :thinking:

Reply:
> Great! So, what would be your plan for this PR? :grin:

I'm going to first open a new PR moving there the logic of the parameters handler, then, when it get merged, I rebase this PR that at that point will just contain the bindings.

> I did not follow this discussion to get the problem, but if there is some limitation in YARP that you think could be addressed, I think we can open an issue upstream.

Thanks @S-Dafarra for the accurate recap. The problem can be seen from another point of view. The string serialization of a list parameter with two string elements and the serialization of a group with a string variable are the same:

```ini
MyVectorOfStrings    (John Doe)

[MY_GROUP]
John                  Doe
```

Are both serialized as:

```
(MyVectorOfString (John Doe))
(MY_GROUP (John Doe))
```

This means that when a Searchable is parsed from its string representation as in the `YarpImplementation::set`, these two cases cannot be identified unambiguously.

@S-Dafarra proposes to keep the parsing as it is and, in this corner case, the new methods that return group and parameter names will both contain the `MyVectorOfString` and `MY_GROUP` names. Instead, I proposed to add them both to the storage (and, therefore, they will show up in both the new methods as well). The difference is that what I propose only affects parsing from the `YarpImplementation::set` method, while maintaining the semantics of `IParameterHandler` correct. I think that for a polymorphic type this is the best way to proceed, also because this limitation can be lifted as soon as YARP will provide a mean to address this corner case.

Reply:
As per https://github.com/dic-iit/bipedal-locomotion-framework/pull/134#issuecomment-725388389, the `IParametersHandler` modifications have been removed from this PR.